Main.hs

```diff
@@ -111,39 +112,51 @@ isOp _                         = False
 
 -- Builtins ---------------------------------------------------------------
 
-data Prim = Nat | List | Unit | Cons | Nil
+data Prim = Unit
+          | Nat | Float | Word64
+          | Char
+          | List | Cons | Nil
   deriving (Show, Eq)
 
 type Builtins = Map QName Prim
 
 getBuiltins :: TCM Builtins
 getBuiltins = Map.fromList . concat <$> mapM getB
-                [ (builtinNat,  Nat)
-                , (builtinList, List)
-                , (builtinUnit, Unit)
-                , (builtinCons, Cons)
-                , (builtinNil,  Nil)
-                ]
+  [ builtinUnit |-> Unit
+  , builtinNat |-> Nat, builtinFloat |-> Float
+  , builtinWord64 |-> Word64
+  , builtinChar |-> Char
+  , builtinList |-> List , builtinCons |-> Cons , builtinNil |-> Nil
+  ]
   where
+    (|->) = (,)
     getB (b, t) = getBuiltin' b >>= \ case
       Nothing          -> return []
       Just (Def q _)   -> return [(q, t)]
       Just (Con c _ _) -> return [(conName c, t)]
       Just _           -> __IMPOSSIBLE__
 
 compilePrim :: Prim -> Hs.QName ()
-compilePrim Nat  = Hs.UnQual () (hsName "Integer")
-compilePrim List = Hs.Special () (Hs.ListCon ())
-compilePrim Unit = Hs.Special () (Hs.UnitCon ())
-compilePrim Cons = Hs.Special () (Hs.Cons ())
-compilePrim Nil  = Hs.Special () (Hs.ListCon ())
+compilePrim = \case
+  Unit   -> special Hs.UnitCon
+  Nat    -> unqual "Integer"
+  Float  -> unqual "Double"
+  Word64 -> qual "Data.Word" "Word64"
```

Comment:
Having this qualified isn't very nice in the generated code. You anyway have to import `Data.Word` even to get the qualified version, so I see no strong reason to not simply use `Word64`.

Reply:
It would be nice if we automatically inserted the necessary imports for built-in types, what do you think?

Reply:
Yeah, why not. We are parsing the user imports, so it should be easy enough to check if it's already there. You probably also want to check that you don't have `Word64` explicitly imported from another module. For checking whether or not it's used I guess the easiest solution is to traverse the generated Haskell code.